### ISSUE TITLE
Error in results[[elt]][[K_index, ll_index]] <- el(res, where = elt) : incompatible types (from NULL to list) in [[ assignment (aka NAs are not supported)

### ISSUE THREAD

**Issue by Rahel14350** [lutsik/MeDeCom#6]
Hi,
I have RRBS data from 44 Lung cancer samples (Ctrl and diseases). There is high heterogeneity beside the tumor cells in the samples, that is why after pre-processing with RnBeads, I submitted the Beta value matrix to MeDeCom. Now, I got this error, Can you please help me. Many thanks in advance,
### >medecom.result<-runMeDeCom(persites_betaValue, Ks=3:10, lambdas=c(0,10^(-5:-1)), NINIT=10, NFOLDS=10, ITERMAX=300, NCORES=9)
[Main:] checking inputs
[Main:] preparing data
[Main:] preparing jobs
[Main:] 2768 factorization runs in total
[Main:] finished all jobs. Creating the object
Error in results[[elt]][[K_index, ll_index]] <- el(res, where = elt) : 
  incompatible types (from NULL to list) in [[ assignment
In addition: Warning messages:
1: In mclapply(rev(concurr_indices), function(index_group) { :
  80 function calls resulted in an error
2: In mclapply(rev(concurr_indices), function(index_group) { :
  8 function calls resulted in an error

**Comment by shmilyfhh** [comment]
I have the sample problem as well.

**Comment by lutsik** [comment]
This error message is unspecific, merely indicating that the worker threads have failed in a multicore setting. To get the actual exception, try rerunning with NCORES=1.

**Comment by nikitaved** [comment]
Could be a problem within a solver. Could you please share the dataset and/or your script?

**Comment by shmilyfhh** [comment]
Since we are working on large dataset, 40 cores took two days to get it done. I will try with one node, and see what is happening. By the way, does MeDeCom takes dataset with NAs? In my situation, it went wrong when my dataset has NAs.

Thanks,
huihui

**Comment by shmilyfhh** [comment]
What I did is just to use the example lines included in your manual. I run into errors with NAs included, but I don’t want to lose too much information, so I first filtered probes with NAs within more than half samples and then replace the remaining NAs with 0s. In order to reduce the computation time, I increase the core number to 40.

Thanks,
huihui

**Comment by lutsik** [comment]
Regardless of decomposition, substituting NAs with zeroes is a bad idea anyway since 0 is one of the two extreme methylation values and will definitely lead to artefacts in your analysis.

I would suggest to reduce the data set to something minimal (several thousands of rows) and try running on one core to identify the source of your problem. 

Did the test data set work fine for you?

**Comment by shmilyfhh** [comment]
Yes, the test data set works just fine both on one core and multiple cores. I agree with what you said, that was a bad decision.

Sure, I could definitely try a subset of probes and go from there.

Thanks,
huihui

**Comment by Rahel14350** [comment]
I also have problem with missing values and it is not even possible to get 1000 rows without NAs to run the test. Is there any solution for the NAs for example replacing them with 0.5? Or should I continue till the end of analysis with a small subset of my data (for example 500 rows)? ...
Many thanks in advance,

**Comment by shmilyfhh** [comment]
What I am doing with NAs is first I substitute them with 0s, and then I masked them as before with NAs after I got the results. Probes with 0.5ish methylation levels are considered as proof showing multiple cell compositions. I don’t think that is a good idea either. Not sure how Lutsik thinks though.

Thanks,
huihui

**Comment by nikitaved** [comment]
The solver does **not** support NAs, that is for sure.

**Comment by Rahel14350** [comment]
For me, also when I replace the NAs value with some number, everything is working on MeDeCom, but I know that this is wrong. At the end there is no solution for NAs? If I continue with only a subset of my methylation values then is the analysis and output correct?

**Comment by nikitaved** [comment]
No, NAs are not supported as of now. Shall we just ignore their contribution in the loss function? We will check how people solve the task of missing values, and, if it is not that much of a work, we will add this feature once we have a bit more time... But I certainly agree it is worth doing.

**Comment by lutsik** [comment]
As a temporary solution, I would recommend looking into the imputation methods in case removing all NAs seems like an overkill.

In fact I started implementing something along these lines a while ago (see argument na.values and its usage in onerun.alternate, which is factorisations.R:445 and below). The idea was to disregard NAs when calculating the loss, and exclude NA-containing rows for the T and A updates. Not sure this was the right way, though.

**Comment by nikitaved** [comment]
`In fact I started implementing something along these lines a while ago (see argument na.values and its usage in onerun.alternate, which is factorisations.R:445 and below). The idea was to disregard NAs when calculating the loss, and exclude NA-containing rows for the T and A updates. Not sure this was the right way, though.`
I think it would not be that much different from simply removing NAs from a dataset (maybe slower convergence but smaller matrices)... I think we could develop an algorithm to solve an adopted version of the matrix completion problem (0-1 constraints) and then apply MeDeCom...

**Comment by Rahel14350** [comment]
Dear Nikitaved and Dear Pavel,
Many thanks for your effort and time to solve this issue. I was also thinking to impute all the NAs using impute package in R (If it is the right solution to do) but you are talking to remove all the probes with NAs (if I am right). In my case removing probes with NAs will remove 90% of probes, is this then correct to continue to the MeDeCom analysis afterwards with only 10% subset of samples? Many tanks in advance,

**Comment by nikitaved** [comment]
Well, the more you use, the better, but much slower =) It also depends on your problem - could be you loose no information after all the filtering. Unfortunately, MeDeCom is not designed to deal with missing data as of now, because the inference of unknown observations is a bit different task.

**Comment by Rahel14350** [comment]
Dear Nikitaved and Dear Pavlo,
I have imputed all the NAs in my Beta value matrix using champ (now I have more than million of sites in 44 samples). Then I did run again MeDeCom but again I got the same error. I am sure that I have no NAs in the matrix. Would you please help me here ...
medecom.result <- runMeDeCom(as.matrix(beta), Ks=2:10, lambdas=c(0,10^(-5:-1)), NINIT=10, NFOLDS=10, ITERMAX=300, NCORES=10)

[Main:] finished all jobs. Creating the object
Error in results[[elt]][[K_index, ll_index]] <- el(res, where = elt) : 
  incompatible types (from NULL to list) in [[ assignment
In addition: Warning message:
In mclapply(rev(concurr_indices), function(index_group) { :
  3 function calls resulted in an error
> warnings()
Warning message:
In mclapply(rev(concurr_indices), function(index_group) { ... :
  3 function calls resulted in an error

**Comment by lutsik** [comment]
Dear Rahel,

As already explained, in a multicore setting the exception pointing at the actual problem is not accessible. Please, rerun with NCORES=1.

Best, 

Pavlo

**Comment by Rahel14350** [comment]
Dear Pavlo, Many thanks for your prompt reply. With this number of cores it took me more than 1 week to get this error. Then it will be slower, but I will do it, when it is the only solution.
Kind regards,
Rahel

**Comment by lutsik** [comment]
First check whether a small subset of your data runs without problems. Take 100 randomly selected rows and run with 1 core.

**Comment by Rahel14350** [comment]
Yes, I already did it and it worked well, that is why I imputed the NAs and did with all the data ...

**Comment by lutsik** [comment]
what is the total number of rows (CpGs)?

**Comment by Rahel14350** [comment]
1282369 CpGs and 44 samples

**Comment by lutsik** [comment]
Applying MeDeCom to over 1 mln CpGs makes little sense since the vast majority of them are not informative. Calculate row-wise variance and make a histogram of that. You will see that most of your rows have zero or very close to zero variance and have no benefit for the deconvolution. Find a reasonable cutoff and select 50-100k rows at most.

**Comment by Rahel14350** [comment]
OK, I am going to do that and continue with a subset of samples. But the last question (Sorry, if it is not related to the question above), Is it possible to extract the methylomes that reflect the underlying biology of constituent cell types after all? I mean when I continue with a subset of samples and the result showed that I have 5 different cell types, Can I only get the methylomes most related to Tumor cells to continue with differential methylome analysis (Please let me know if I am wrong and did not get the idea of MeDeCom)?

**Comment by lutsik** [comment]
If the method works as expected and reveals 5 underlying cell types, for each of them you will get proportions in each sample as well as an "average" methylation signature (but only in the rows that you have used for the deconvolution). For the differential methylation analysis you should include the obtained proportions as covariates.

**Comment by Rahel14350** [comment]
I am so grateful for your prompt reply. therefore, I will only continue with a subset of my data. What if I make several subset of CpGs and at the end merge all the proportions for each samples in tumor cells? I hope this is also correct.
